### DISCUSSION TITLE
Custom OAuth SignIn Page Not Redirecting After Authentication

### DISCUSSION THREAD

**Discussion by lsbyerley** [nextauthjs/next-auth#4366] [Help] [answered]
Hello.. I am following the documentation on how to setup a custom oauth signin page and after authenticating, the signin page does not redirect the user to the callbackUrl. Is there anything missing from the example here:

https://next-auth.js.org/configuration/pages#oauth-sign-in

I have copy/pasted the same code from the docs and i have placed the file at `pages/auth/signin.js` 

After authtication, the url is `http://localhost:3000/auth/signin?callbackUrl=http%3A%2F%2Flocalhost%3A3000%2F` 

```js

import { getProviders, signIn } from 'next-auth/react';

export default function SignIn({ providers }) {
  return (
    <div className='max-w-2xl mx-auto'>
      <div className='flex items-center justify-center px-8 py-8 mt-4 rounded bg-content bg-base-200'>
        {Object.values(providers).map((provider) => (
          <div key={provider.name}>
            <button className='btn' onClick={() => signIn(provider.id)}>
              Sign in with {provider.name}
            </button>
          </div>
        ))}
      </div>
    </div>
  );
}

// This is the recommended way for Next.js 9.3 or newer
export async function getServerSideProps(context) {
  const providers = await getProviders();
  return {
    props: { providers },
  };
}
```

**Comment by ThangHuuVu** [comment]
The code looks good to me, so I suspect this is a configuration error. Could you provide the rest of the code?

**Comment by lsbyerley** [comment]
I'm not entirely sure what else to show other than the NextAuth route with the custom page option. 

```js
import NextAuth from 'next-auth';
import GoogleProvider from 'next-auth/providers/google';

export default NextAuth({
  pages: {
    signIn: '/auth/signin',
  },
  // Configure one or more authentication providers
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
    }),
  ],
  debug: process.env.NEXTAUTH_DEBUG,
});
```

**Comment by ThangHuuVu** [comment, marked as answer]
oh, it could be because of the custom sign-in page. Could you try something like this:
```js
export default function SignIn({ providers }) {
  const { callbackUrl }  = useRouter().query
  return (
    <div className='max-w-2xl mx-auto'>
      <div className='flex items-center justify-center px-8 py-8 mt-4 rounded bg-content bg-base-200'>
        {Object.values(providers).map((provider) => (
          <div key={provider.name}>
            <button className='btn' onClick={() => signIn(provider.id, { callbackUrl })}>
              Sign in with {provider.name}
            </button>
          </div>
        ))}
      </div>
    </div>
  );
}
```
It seems like the documentation is missing this piece about custom sign-in page 🤔 cc @ndom91

**Reply by lsbyerley** [reply]
Yes, this was the fix and is not documented at the moment

**Reply by forntoh** [reply]
`callbackUrl` is undefined, how do I fix that?

**Reply by lsbyerley** [reply]
@forntoh you can pass it to the signin function as an option parameter if it's undefined

`<button className='btn' onClick={() => signIn(provider.id, { callbackUrl: '/' })}>`

**Reply by forntoh** [reply]
> @forntoh you can pass it to the signin function as an option parameter if it's undefined
> 
> 
> 
> `<button className='btn' onClick={() => signIn(provider.id, { callbackUrl: '/' })}>`

Are relative paths allowed too?

**Reply by lsbyerley** [reply]
https://next-auth.js.org/getting-started/client#specifying-a-callbackurl

**Reply by tiendq** [reply]
> Yes, this was the fix and is not documented at the moment

The example code is still not fixed, and both relative and absolute path for `callbackUrl` don't work, it still redirect to homepage.

Update: `callbackUrl` works, I forgot to remove `redirect` callback added for testing purpose.

> The URL must be considered valid by the [redirect callback handler](https://next-auth.js.org/configuration/callbacks#redirect-callback). By default it requires the URL to be an absolute URL at the same host name, or a relative url starting with a slash. If it does not match it will redirect to the homepage. You can define your own [redirect callback](https://next-auth.js.org/configuration/callbacks#redirect-callback) to allow other URLs.
